lockfile/src/spdx.rs

```diff
@@ -76,62 +79,84 @@ fn type_from_url(url: &str) -> anyhow::Result<PackageType> {
+
+    let package_type = PackageType::from_str(purl.ty())
+        .or_else(|_| type_from_url(&pkg_info.download_location))
+        .context(UnknownEcosystem)?;
+
+    let name = match (package_type, purl.namespace()) {
+        (PackageType::Maven, Some(ns)) => format!("{}:{}", ns, purl.name()),
+        (PackageType::Npm | PackageType::Golang, Some(ns)) => format!("{}/{}", ns, purl.name()),
+        _ => purl.name().into(),
+    };
+
+    let pkg_version = pkg_info
+        .version_info
+        .as_ref()
+        .ok_or(purl.version())
+        .map_err(|_| anyhow!("No version found for `{}`", pkg_info.name))?;
+
+    let version = purl
+        .qualifiers()
+        .iter()
+        .find_map(|(key, value)| match key.as_ref() {
+            "repository_url" => Some(PackageVersion::ThirdParty(ThirdPartyVersion {
+                version: pkg_version.clone(),
+                registry: value.to_string(),
+            })),
+            "download_url" => Some(PackageVersion::DownloadUrl(value.to_string())),
+            "vcs_url" => {
+                if value.as_ref().starts_with("git+") {
+                    Some(PackageVersion::Git(value.to_string()))
+                } else {
+                    None
+                }
+            },
+            _ => None,
+        })
+        .unwrap_or(PackageVersion::FirstParty(pkg_version.into()));
+
+    Ok(Package { name, version, package_type })
+}
+
+fn from_locator(registry: &str, locator: &str) -> anyhow::Result<Package> {
+    lazy_static! {
+        static ref RE: Regex = Regex::new(r"^(([^:]*:)*[^@]+)@([^:]+)$").unwrap();
+    }
+
+    let package_type = PackageType::from_str(registry).map_err(|_| UnknownEcosystem)?;
+    let captures = RE.captures(locator).ok_or(anyhow!("Invalid locator: {}", locator))?;
+    let name =
+        decode(captures.get(1).ok_or(anyhow!("Missing package name: {}", locator))?.as_str())?;
+    let version = captures.get(3).ok_or(anyhow!("Missing package version: {}", locator))?.as_str();
+
+    Ok(Package {
+        name: name.into(),
+        version: PackageVersion::FirstParty(version.into()),
+        package_type,
+    })
+}
+
 impl TryFrom<&PackageInformation> for Package {
     type Error = anyhow::Error;
 
     fn try_from(pkg_info: &PackageInformation) -> anyhow::Result<Self> {
-        let pkg_url = pkg_info
+        pkg_info
             .external_refs
             .iter()
             .find_map(|external| match external.reference_category {
                 ReferenceCategory::PackageManager => {
                     if external.reference_type == "purl" {
-                        Some(&external.reference_locator)
+                        Some(from_purl(&external.reference_locator, pkg_info))
                     } else {
-                        None
+                        Some(from_locator(&external.reference_type, &external.reference_locator))
                     }
                 },
                 _ => None,
             })
```

Comment:
Is it possible to have multiple `PACKAGE-MANAGER` locators? If so, this will only ever attempt to parse the first one. What if we can't parse the first one, but we could parse one of the other locators?

Reply:
I don't think anything prevents it, so it's technically possible, but you should only have one to identify where the package actually came from?